File: plugins/a2a-protocol/skills/a2a-executor-patterns/templates/basic-executor.py
"""
Basic A2A Executor - Synchronous Pattern

Use for: Simple, fast tasks with immediate results
Examples: Validation, quick transformations, data formatting
"""

from dataclasses import dataclass, field
from typing import Any, Dict, Optional
from datetime import datetime
import json


# Task and Result Classes
@dataclass
class A2ATask:
    """A2A Task definition"""
    id: str
    type: str
    parameters: Dict[str, Any]
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class A2AResult:
    """A2A Result definition"""
    task_id: str
    status: str  # 'completed' or 'failed'
    result: Optional[Any] = None
    error: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


# Error Classes
class ValidationError(Exception):
    """Raised when task validation fails"""
    pass


class ExecutionError(Exception):
    """Raised when task execution fails"""
    pass


# Task Validation
def validate_task(task: A2ATask) -> None:
    """Validate task structure and required fields"""
    if not task.id:
        raise ValidationError('Task ID is required')

    if not task.type:
        raise ValidationError('Task type is required')

    if not isinstance(task.parameters, dict):
        raise ValidationError('Task parameters must be a dictionary')


# Main Executor Function
async def execute_task(task: A2ATask) -> A2AResult:
    """Execute A2A task and return result"""
    try:
        # Step 1: Validate input
        validate_task(task)

        # Step 2: Process task based on type
        result = None

        if task.type == 'validate-data':
            result = await validate_data(task.parameters)

        elif task.type == 'transform-data':
            result = await transform_data(task.parameters)

        elif task.type == 'compute-result':
            result = await compute_result(task.parameters)

        else:
            raise ExecutionError(f'Unsupported task type: {task.type}')

        # Step 3: Return successful result
        return A2AResult(
            task_id=task.id,
            status='completed',
            result=result,
            metadata={
                'executed_at': datetime.utcnow().isoformat(),
                'execution_time': 0  # Add timing if needed
            }
        )

    except Exception as error:
        # Error handling
        print(f'Task execution failed: {task.id} - {str(error)}')

        return A2AResult(
            task_id=task.id,
            status='failed',
            error=str(error),
            metadata={
                'failed_at': datetime.utcnow().isoformat()
            }
        )


# Example Task Handlers
async def validate_data(parameters: Dict[str, Any]) -> Any:
    """Implement validation logic"""
    data = parameters.get('data')
    schema = parameters.get('schema')

    # Example: simple validation
    if not data:
        raise ValidationError('Data is required')

    return {
        'valid': True,
        'data': data
    }


async def transform_data(parameters: Dict[str, Any]) -> Any:
    """Implement transformation logic"""
    input_data = parameters.get('input')
    transform_type = parameters.get('transformType')

    # Example: simple transformation
    return {
        'transformed': input_data,
        'type': transform_type
    }


async def compute_result(parameters: Dict[str, Any]) -> Any:
    """Implement computation logic"""
    values = parameters.get('values')
    operation = parameters.get('operation')

    # Example: simple computation
    return {
        'result': values,
        'operation': operation
    }


# Example Usage
if __name__ == '__main__':
    import asyncio

    async def main():
        example_task = A2ATask(
            id='task-001',
            type='validate-data',
            parameters={
                'data': {'name': 'test', 'value': 42},
                'schema': {}
            }
        )

        result = await execute_task(example_task)
        print('Result:', json.dumps({
            'task_id': result.task_id,
            'status': result.status,
            'result': result.result,
            'error': result.error,
            'metadata': result.metadata
        }, indent=2))

    asyncio.run(main())

File: plugins/a2a-protocol/skills/a2a-executor-patterns/templates/basic-executor.ts
/**
 * Basic A2A Executor - Synchronous Pattern
 *
 * Use for: Simple, fast tasks with immediate results
 * Examples: Validation, quick transformations, data formatting
 */

// Task and Result Types
interface A2ATask {
  id: string
  type: string
  parameters: Record<string, any>
  metadata?: Record<string, any>
}

interface A2AResult {
  taskId: string
  status: 'completed' | 'failed'
  result?: any
  error?: string
  metadata?: Record<string, any>
}

// Error Classes
class ValidationError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ValidationError'
  }
}

class ExecutionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'ExecutionError'
  }
}

// Task Validation
function validateTask(task: A2ATask): void {
  if (!task.id) {
    throw new ValidationError('Task ID is required')
  }

  if (!task.type) {
    throw new ValidationError('Task type is required')
  }

  if (!task.parameters || typeof task.parameters !== 'object') {
    throw new ValidationError('Task parameters must be an object')
  }
}

// Main Executor Function
async function executeTask(task: A2ATask): Promise<A2AResult> {
  try {
    // Step 1: Validate input
    validateTask(task)

    // Step 2: Process task based on type
    let result: any

    switch (task.type) {
      case 'validate-data':
        result = await validateData(task.parameters)
        break

      case 'transform-data':
        result = await transformData(task.parameters)
        break

      case 'compute-result':
        result = await computeResult(task.parameters)
        break

      default:
        throw new ExecutionError(`Unsupported task type: ${task.type}`)
    }

    // Step 3: Return successful result
    return {
      taskId: task.id,
      status: 'completed',
      result,
      metadata: {
        executedAt: new Date().toISOString(),
        executionTime: 0 // Add timing if needed
      }
    }
  } catch (error) {
    // Error handling
    console.error('Task execution failed:', {
      taskId: task.id,
      error: error instanceof Error ? error.message : 'Unknown error'
    })

    return {
      taskId: task.id,
      status: 'failed',
      error: error instanceof Error ? error.message : 'Unknown error',
      metadata: {
        failedAt: new Date().toISOString()
      }
    }
  }
}

// Example Task Handlers
async function validateData(parameters: Record<string, any>): Promise<any> {
  // Implement validation logic
  const { data, schema } = parameters

  // Example: simple validation
  if (!data) {
    throw new ValidationError('Data is required')
  }

  return {
    valid: true,
    data
  }
}

async function transformData(parameters: Record<string, any>): Promise<any> {
  // Implement transformation logic
  const { input, transformType } = parameters

  // Example: simple transformation
  return {
    transformed: input,
    type: transformType
  }
}

async function computeResult(parameters: Record<string, any>): Promise<any> {
  // Implement computation logic
  const { values, operation } = parameters

  // Example: simple computation
  return {
    result: values,
    operation
  }
}

// Export
export {
  A2ATask,
  A2AResult,
  executeTask,
  validateTask,
  ValidationError,
  ExecutionError
}

// Example Usage
if (require.main === module) {
  const exampleTask: A2ATask = {
    id: 'task-001',
    type: 'validate-data',
    parameters: {
      data: { name: 'test', value: 42 },
      schema: {}
    }
  }

  executeTask(exampleTask)
    .then(result => {
      console.log('Result:', JSON.stringify(result, null, 2))
    })
    .catch(error => {
      console.error('Error:', error)
    })
}
